README.md
# Calen

STDOUTにカレンダーを出力する。

## Installation

    $ git clone https://github.com/gpsoft/calen.git
    $ boot build

## Usage

    $ java -jar target/calen-1.0.0-standalone.jar ARGS

    ARGS: YEAR [MONTH or NUMOFCOLUMNS(in negative integer)]

## Examples

    $ java -jar target/calen-1.0.0-standalone.jar 2017 5
    $ java -jar target/calen-1.0.0-standalone.jar 2017 -2

ロケール指定が必要かも。

    $ LC_ALL=ja_JP.utf8 \
      java -jar target/calen-1.0.0-standalone.jar 2017 5
    2017年5月           
    日 月 火 水 木 金 土
        1  2  3  4  5  6
     7  8  9 10 11 12 13
    14 15 16 17 18 19 20
    21 22 23 24 25 26 27
    28 29 30 31         


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-05-03
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-05-03
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/calen/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/calen/compare/0.1.0...0.1.1


test/calen/core_test.clj
(ns calen.core-test
  (:require [clojure.test :refer :all]
            [calen.core :refer :all]))

(deftest not-leap-test
  (is (not (leap? 2017))))
(deftest normal-leap-test
  (is (leap? 2016)))
(deftest special-not-leap-test
  (is (not (leap? 2100))))
(deftest special-leap-test
  (is (leap? 2000)))

(deftest dow-2017-test
  (is (= 0 (dow-of-nyd 2017))))
(deftest dow-2018-test
  (is (= 1 (dow-of-nyd 2018))))

(deftest dow-of-bom-20170101-test
  (is (= 0 (bom-dow 2017 1))))
(deftest dow-of-bom-20170201-test
  (is (= 3 (bom-dow 2017 2))))
(deftest dow-of-bom-20171201-test
  (is (= 5 (bom-dow 2017 12))))

(deftest dow-ad1-test
  (is (= 1 (dow [1 1 1]))))
(deftest dow-20170503-test
  (is (= 3 (dow [2017 5 3]))))

(comment
  (require '[calen.core :refer :all] :reload-all)
  (num-days-list 2017)
  (num-days-before-list 2017)
  (dow-of-nyd 2017)
  (num-days-before 2)
  (num-days-before 2017)
  (num-days-before' 2)
  (num-days-before' 2017)
  (bom-dow 2017 2)
  (bom-dow-list 2017)
  (dow-name (dow [2017 5 3]))
  (dow-name (dow [2017 5 3]))
  )


src/calen/fig.clj
(ns calen.fig)

;; データ構造
;; fig: {:width 3, :height 2, :body ["a" "b"]}

(defn ->fig [w h b] {:width w, :height h, :body b})
(defn space [len] (apply str (repeat len " ")))
(defn vspace [n] (repeat n ""))
(defn wcount [s] (count (.getBytes s "Shift_JIS")))

(defn vertical
  "図形を上下に連結"
  [gap fig1 fig2]
  (let [w1 (:width fig1)
        w2 (:width fig2)
        h1 (:height fig1)
        h2 (:height fig2)
        b1 (:body fig1)
        b2 (:body fig2)]
    {:width (max w1 w2)
     :height (+ h1 gap h2)
     :body (concat b1 (vspace gap) b2)}))

(defn horizon
  "図形を水平に連結"
  [gap fig1 fig2]
  (let [w1 (:width fig1)
        w2 (:width fig2)
        h1 (:height fig1)
        h2 (:height fig2)
        b1 (concat (:body fig1) (vspace h2))
        b2 (concat (:body fig2) (vspace h1))
        h (max h1 h2)
        concat-with-gap
        (fn [l r]
          (str l (space (- w1 (wcount l))) (space gap) r))]
    {:width (+ w1 gap w2)
     :height h
     :body (->> (map concat-with-gap b1 b2)
                (take h))}))

(defn grid
  "複数の図形をグリッド上に並べて1つの図形にする"
  [figs numcols hgap vgap]
  (let [rows (partition numcols numcols nil figs)]
    (reduce (partial vertical vgap)
            (for [figs rows]
              (reduce (partial horizon hgap) figs)))))

(defn prn-fig
  "図形を印字"
  [{:keys [width height body]}]
  (doall
    (for [line body]
      (let [len (wcount line)
            padding (space (- width len))]
        (println (str line padding)))))
  nil)

(comment
  (prn-fig {:width 3, :height 2, :body ["a" "b"]})
  (-> (vertical
        2
        {:width 3, :height 2, :body ["a" "b"]}
        {:width 4, :height 3, :body ["a" "b" "c"]})
      (prn-fig))
  (-> (horizon
        2
        {:width 3, :height 2, :body ["a" "b"]}
        {:width 4, :height 3, :body ["a" "b" "c"]})
      (prn-fig))
  (grid [{:width 3, :height 2, :body ["a" "b"]}
         {:width 3, :height 2, :body ["a" "b"]}
         {:width 3, :height 2, :body ["a" "b"]}] 2 3 1)
  (reduce (partial horizon 3) [{:width 3, :height 2, :body ["a" "b"]}
         {:width 3, :height 2, :body ["a" "b"]}
         {:width 3, :height 2, :body ["a" "b"]}])
  )


src/calen/core.clj
(ns calen.core
  (:require [calen.fig :as fig]))

;; 略語とデータ構造
;;
;; nyd: new year's day, 元日
;; bom: beginning of month, ついたち
;; dow: day of week, 曜日
;;
;; date: [year month day]
;; dow: 0-6, 0:Sunday, 1:Monday, ...,6:Saturday
;; ad1: [1 1 1], nyd of AD 1


;; 例えば2017年5月1日の曜日を得るには…
;; 　・元日の曜日
;;   ・各月の日数
;;   ・元日から各月の1日までの日数
;; などから計算

;; [31 28 31 30 31 30 31 31 30 31 30 31]  ...各月の日数
;; ↓
;; [0 31 59 90 120 151 ...]  ...元日からの日数
;; ↓
;; (元日の曜日 + 120) % 7



(def ad1 [1 1 1])  ;; 0001-01-01
(def dow-ad1 1)    ;; 0001-01-01は月曜
(def dow-names ["日" "月" "火" "水" "木" "金" "土"])

(defn- y-of [[y]] y)
(defn- m-of [[_ m]] m)
(defn- d-of [[_ _ d]] d)

(defn- dow-add
  "ある曜日(dow-base)の日から
  何日(num-days)か後の曜日"
  [dow-base num-days]
  (-> (+ dow-base num-days)
      (mod 7)))


(defn leap?
  "うるう年か?
  true for 2016, 2020, 2000, 2400
  false for 2017, 2100, 2200
  "
  [y]
  (if (zero? (mod y 100))
    (zero? (mod y 400))
    (zero? (mod y 4))))

(defn num-days-before
  "ad1から前年まで(ad1を含めて)何日あった?"
  [y]
  (->> (range 1 y)
         (map #(if (leap? %) 366 365))
         (reduce +)))
(defn num-days-before'
  "num-days-beforeの高速バージョン?"
  [y]
  (let [x (dec y)
        num-leap-years (+ (quot x 4)
                          (- (quot x 100))
                          (quot x 400))]
    (-> (* x 365)
        (+ num-leap-years))))

(defn dow-of-nyd
  "元日(new year's day)の曜日"
  [y]
  (let [ndb (num-days-before y)]
    (dow-add dow-ad1 ndb)))

(defn num-days-list
  "各月の日数のリスト
  [31 28 31 ...]
  "
  [y]
  [31 (if (leap? y) 29 28) 31 30
   31 30 31 31 30 31 30 31])

(defn num-days-before-list
  "年初から前月までの日数のリスト
  (0 31 59 90 ...)
  "
  [y]
  (let [ndl (num-days-list y)]
    (->> (reductions + 0 ndl)
         (take 12))))

(defn bom-dow-list
  "ついたちの曜日のリスト"
  [y]
  (let [downyd (dow-of-nyd y)]
    (->> (num-days-before-list y)
        (map #(dow-add downyd %)))))

(defn num-days
  "月の日数"
  [y m]
  (nth (num-days-list y) (dec m)))

(defn bom-dow
  "ついたちの曜日"
  [y m]
  (nth (bom-dow-list y) (dec m)))

(defn dow
  "曜日"
  [[y m d :as date]]
  (let [bomdow (bom-dow y m)]
    (dow-add bomdow (dec d))))

(defn dow-name
  "曜日名"
  [dow]
  (nth dow-names dow))



(defn render-ym
  "年月ラベル"
  [y m]
  (str y "年" m "月"))

(defn render-dows
  "曜日ラベル"
  [hgap]
  (clojure.string/join (fig/space hgap) dow-names))

(defn day->fig
  "1日分を図形化する"
  [d]
  (let [strd (if (pos? d) (format "%2d" d) "")]
    (fig/->fig 2 1 [strd])))

(defn month->fig
  "ひと月分を図形化する"
  [y m]
  (let [hgap 1
        vgap 0
        heading-dow (render-dows hgap)
        headings (fig/->fig
                   (count heading-dow) 2
                   [(render-ym y m) heading-dow])
        bomdow (bom-dow y m)
        nd (num-days y m)
        ds (map inc (range (- bomdow) nd))]
    (fig/vertical
      vgap
      headings
      (fig/grid (map day->fig ds) 7 hgap vgap))))

(defn year->fig
  "1年分を図形化する"
  [y numcols]
  (let [hgap 3
        vgap 1]
    (fig/grid
      (map #(month->fig y %) (range 1 13))
      numcols hgap vgap)))

(comment
  (fig/prn-fig (month->fig 2017 5))
  (fig/prn-fig (year->fig 2018 4))
  )
